libs/meta/include/meta/param_pack.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <cstddef>
+#include <tuple>
+#include <utility>
+
+namespace fetch {
+namespace meta {
+
+namespace internal {
+
+template <template <typename...> class Source, typename SourceInstantiation,
```

Comment:
I don't know what you mean here. The intent was to allow clients to easily create typesafe aliases for whichever template they happen to be working with - see for example `UnpackTuple` in this changeset. As it is, its definition is one line and it will only accept instantiations of `std::tuple`, but without putting the burden of implicitly specialising on `std::tuple` on the client. Not to mention that they would have to go through the usual `struct-struct-using` dance to do that, as you cannot specialise type alias templates. How would you have done this?

Reply:
What I mean is as it's written it is invoked as `ConveyTypeParameterPack<Source, Source<int, char>, Dest>`. Note the `Source` typed twice. The first template parameter is excessive as it can be deduced from the second one.

Reply:
I explained above why I chose to write it the way I did. `ConveyTypeParameterPack` is a low-level abstraction that will likely be (or currently is, anyway) used via semantic aliasing to improve code readability - `UnpackTuple` being an example.

Let's compare the current API in action and what it could look like if it relied on deduction.

* Current definition (3 template arguments):
```
template <template <typename...> class, typename, template <typename...> class>
struct ConveyTypeParameterPack;
template <template <typename...> class Source, template <typename...> class Destination,
          typename... Args>
struct ConveyTypeParameterPack<Source, Source<Args...>, Destination>
{
  using type = Destination<Args...>;
};
```

Current usage (direct instantiation):
```
ConveyTypeParameterPack<Source, Source<T1, T2>, Destination>;
```

Current usage (type-safe alias):
```
template <typename Tuple, template <typename...> class Destination>
using UnpackTuple = ConveyTypeParameterPack<std::tuple, Tuple, Destination>;
```

* Remove redundant argument and rely on deduction. Definition:
```
template <typename, template <typename...> class>
struct ConveyTypeParameterPack;
template <template <typename...> class Destination, template <typename...> class Source,
          typename... Args>
struct ConveyTypeParameterPack<Source<Args...>, Destination>
{
  using type = Destination<Args...>;
};
```

Direct usage indeed becomes simpler:
```
ConveyTypeParameterPack<Source<T1, T2>, Destination>;
```

Definition of `UnpackTuple` with type-safety equivalent to previous one:
```
template <typename Tuple, template <typename...> class Destination>
struct UnpackTupleImpl;
template <template <typename...> class Destination, typename... Args>
struct UnpackTupleImpl<std::tuple<Args...>, Destination>
{
  using type = ConveyTypeParameterPack<std::tuple<Args...>, Destination>;
};
template <typename Tuple, template <typename...> class Destination>
using UnpackTuple = typename UnpackTupleImpl<Tuple, Destination>::type;
```

Can you think of a more parsimonious way to define `UnpackTuple` while maintaining specificity for `std::tuple` as the source template? If not, then I think my point is clear - at this early stage, the usage patterns of `ConveyTypeParameterPack` do not justify bloating one usage scenario to four times the size, just so the other can save a single word. Apart from the obvious maintenance considerations, this kind of unnecessary boilerplate is precisely what gives metaprogramming a bad name with so many developers.

Reply:
If you want it to be `std::tuple`-specific why not keep a single partial instantiation?

Reply:
I'm not going to guess what you men this time. What partial instantiation would that be? I posed a question in my previous comment and you ignored it. Kindly post an example like I did and we can compare the pros and cons of your solution. Otherwise there is no discussion to be had here.

Reply:
Actually, you can keep both:
`template <typename, template <typename...> class>
struct ConveyTypeParameterPack;

template <template <typename...> class Destination, template <typename...> class Source,
          typename... Args>
struct ConveyTypeParameterPack<Source<Args...>, Destination>
{
  using type = Destination<Args...>;
};

template <template<class...> class, typename, template <typename...> class>
struct CheckedConveyTypeParameterPack;

template <template<class...> class Source, template<class...> class Dest, class... Args>
struct CheckedConveyTypeParameterPack<Source, Source<Args...>, Dest>
  : ConveyTypeParameterPack<Source<Args...>, Dest> {};`
But whatever.

Reply:
Thank you for posting the example. I think they make the point very well: removing that first argument leads to pointless code bloat. It even fails on its own terms: you object to typing `Source` twice, but typing an extra `Checked` suffix every time is somehow ok? I'm glad we reached an agreement in the end.